src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -475,12 +479,58 @@ ReturnRowsFromTuplestore(uint64 tupleCount, TupleDesc tupleDescriptor,
  * connection so that we receive results a row at a time.
  */
 static bool
-SendQueryInSingleRowMode(PGconn *connection, char *query)
+SendQueryInSingleRowMode(PGconn *connection, char *query, ParamListInfo paramListInfo)
 {
 	int querySent = 0;
 	int singleRowMode = 0;
 
-	querySent = PQsendQuery(connection, query);
+	if (paramListInfo != NULL)
+	{
+		int parameterIndex = 0;
+		int parameterCount = paramListInfo->numParams;
+		Oid *parameterTypes = palloc(parameterCount * sizeof(Oid));
+		const char **parameterValues = palloc(parameterCount * sizeof(char *));
+
+		/* get parameter types and values */
+		for (parameterIndex = 0; parameterIndex < parameterCount; parameterIndex++)
+		{
+			ParamExternData *parameterData = &paramListInfo->params[parameterIndex];
+			Oid typeOutputFunctionId = InvalidOid;
+			bool variableLengthType = false;
+
+			/*
+			 * Use text oid for data types where the oid values can be different
+			 * on the master and worker nodes.
+			 */
+			if (parameterData->ptype >= FirstNormalObjectId)
```

Comment:
> Might be worthwhile to generalize the function from cab03a6 instead.

This is a bit different than the COPY version. Here, for any of user-defined types, we can't use the original oid of the type because it will be different on master and worker nodes. For example for hll type, COPY version returns false and continues using the text format, but here we need to change the oid.